test/e2e/storage/static_pods.go

```diff
@@ -0,0 +1,184 @@
+	e2evolume "k8s.io/kubernetes/test/e2e/framework/volume"
+	"k8s.io/kubernetes/test/e2e/storage/drivers"
+	storageframework "k8s.io/kubernetes/test/e2e/storage/framework"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+	admissionapi "k8s.io/pod-security-admission/api"
+)
+
+// Tests for static Pods + CSI volume reconstruction.
+//
+// NOTE: these tests *require* the cluster under test to be Kubernetes In Docker (kind)!
+// Kind runs its API server as a static Pod, and we leverage it here
+// to test kubelet starting without the API server.
+var _ = utils.SIGDescribe("StaticPods [Feature:Kind]", func() {
+	f := framework.NewDefaultFramework("static-pods-csi")
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelPrivileged
+
+	// global vars for the ginkgo.BeforeEach() / It()'s below
+	var (
+		driver     storageframework.DynamicPVTestDriver
+		testConfig *storageframework.PerTestConfig
+	)
+
+	ginkgo.BeforeEach(func(ctx context.Context) {
+		// Install a CSI driver. Using csi-driver-hospath here.
+		driver = drivers.InitHostPathCSIDriver().(storageframework.DynamicPVTestDriver)
+		testConfig = driver.PrepareTest(ctx, f)
+	})
+
+	// Test https://github.com/kubernetes/kubernetes/issues/117745
+	// I.e. kubelet starts and it must start the API server as a static pod,
+	// while there is a CSI volume mounted by the previous kubelet.
+	ginkgo.It("should run after kubelet stopped with CSI volume mounted [Disruptive][Serial]", func(ctx context.Context) {
+
+		ginkgo.By("Provision a new CSI volume")
+		res := storageframework.CreateVolumeResource(ctx, driver, testConfig, storageframework.DefaultFsDynamicPV, e2evolume.SizeRange{Min: "1", Max: "1Gi"})
+		ginkgo.DeferCleanup(func(ctx context.Context) {
+			res.CleanupResource(ctx)
+		})
+
+		ginkgo.By("Run a pod with the volume")
+		podConfig := &e2epod.Config{
+			NS:            f.Namespace.Name,
+			PVCs:          []*v1.PersistentVolumeClaim{res.Pvc},
+			NodeSelection: testConfig.ClientNodeSelection,
+		}
+		pod, err := e2epod.CreateSecPod(ctx, f.ClientSet, podConfig, f.Timeouts.PodStart)
+		framework.ExpectNoError(err, "Creating a pod with a CSI volume")
+		ginkgo.DeferCleanup(func(ctx context.Context) {
+			if pod != nil {
+				err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Delete(ctx, pod.Name, metav1.DeleteOptions{})
+				framework.ExpectNoError(err, "Deleting the pod with the CSI volume")
+			}
+		})
+
+		ginkgo.By("Stop the control plane by removing the static pod manifests")
+		err = stopControlPlane(ctx)
+		framework.ExpectNoError(err, "Stopping the control plane")
+
+		ginkgo.By("Wait for the API server to disappear")
+		gomega.Eventually(ctx, func(ctx context.Context) error {
+			return checkAPIServer(ctx, false, f.ClientSet)
+		}).WithTimeout(time.Minute).Should(gomega.Succeed())
+
+		ginkgo.By("Stop kubelet")
+		err = stopKindKubelet(ctx)
+		framework.ExpectNoError(err, "Stopping kubelet")
+
+		ginkgo.By("Re-enable control plane")
+		err = enableControlPlane(ctx)
```

Comment:
There's probably no good way to restore the cluster if this test fails somewhere in the middle?

Reply:
No, I could only delete + re-create the cluster. It's a really disruptive test.